File: jena-patch/src/test/scala/eu/ostrzyciel/jelly/convert/jena/patch/JenaChangesCollector.scala
package eu.ostrzyciel.jelly.convert.jena.patch

import eu.ostrzyciel.jelly.core.proto.v1.patch.PatchStatementType
import org.apache.jena.graph.Node
import org.apache.jena.rdfpatch.RDFChanges
import org.apache.jena.sparql.core.Quad

import scala.collection.mutable.ListBuffer

object JenaChangesCollector:
  sealed trait JenaChangesItem:
    def apply(c: RDFChanges): Unit
  final case class Header(field: String, value: Node) extends JenaChangesItem:
    def apply(c: RDFChanges): Unit = c.header(field, value)
  final case class Add(g: Node, s: Node, p: Node, o: Node) extends JenaChangesItem:
    def apply(c: RDFChanges): Unit = c.add(g, s, p, o)
  final case class Delete(g: Node, s: Node, p: Node, o: Node) extends JenaChangesItem:
    def apply(c: RDFChanges): Unit = c.delete(g, s, p, o)
  final case class AddPrefix(gn: Node, prefix: String, uriStr: String) extends JenaChangesItem:
    def apply(c: RDFChanges): Unit = c.addPrefix(gn, prefix, uriStr)
  final case class DeletePrefix(gn: Node, prefix: String) extends JenaChangesItem:
    def apply(c: RDFChanges): Unit = c.deletePrefix(gn, prefix)
  case object TxnBegin extends JenaChangesItem:
    def apply(c: RDFChanges): Unit = c.txnBegin()
  case object TxnCommit extends JenaChangesItem:
    def apply(c: RDFChanges): Unit = c.txnCommit()
  case object TxnAbort extends JenaChangesItem:
    def apply(c: RDFChanges): Unit = c.txnAbort()
  case object Segment extends JenaChangesItem:
    def apply(c: RDFChanges): Unit = c.segment()

/**
 * Helper class to collect and replay items from Jena RDFChanges streams.
 * @param stType The statement type of the Jelly-Patch stream.
 */
final class JenaChangesCollector(stType: PatchStatementType) extends RDFChanges:
  import JenaChangesCollector.*

  private val changes: ListBuffer[JenaChangesItem] = ListBuffer.empty

  def getChanges: Seq[JenaChangesItem] =
    changes.toSeq

  def size: Int =
    changes.size

  def replay(c: RDFChanges, callStartFinish: Boolean): Unit =
    if callStartFinish then c.start()
    getChanges.foreach(_.apply(c))
    if callStartFinish then c.finish()

  def header(field: String, value: Node): Unit =
    changes += Header(field, value)

  def add(g: Node, s: Node, p: Node, o: Node): Unit =
    changes += Add(coerceGraph(g), s, p, o)

  def delete(g: Node, s: Node, p: Node, o: Node): Unit =
    changes += Delete(coerceGraph(g), s, p, o)

  private def coerceGraph(g: Node): Node =
    if g == null && stType.isQuads then Quad.defaultGraphNodeGenerated
    else if stType.isTriples then null
    else g

  def addPrefix(gn: Node, prefix: String, uriStr: String): Unit =
    changes += AddPrefix(gn, prefix, uriStr)

  def deletePrefix(gn: Node, prefix: String): Unit =
    changes += DeletePrefix(gn, prefix)

  def txnBegin(): Unit =
    changes += TxnBegin

  def txnCommit(): Unit =
    changes += TxnCommit

  def txnAbort(): Unit =
    changes += TxnAbort

  def segment(): Unit =
    changes += Segment

  def start(): Unit = ()

  def finish(): Unit = ()

File: jena-patch/src/main/java/eu/neverblink/jelly/convert/jena/patch/JenaChangesCollector.java
package eu.neverblink.jelly.convert.jena.patch;

import eu.neverblink.jelly.core.ExperimentalApi;
import eu.neverblink.jelly.core.InternalApi;
import eu.neverblink.jelly.core.proto.v1.patch.PatchStatementType;
import java.util.ArrayList;
import java.util.List;
import org.apache.jena.graph.Node;
import org.apache.jena.rdfpatch.RDFChanges;
import org.apache.jena.sparql.core.Quad;

/**
 * A collector for Jena RDFChanges operations that can be replayed later.
 * <p>
 * This class collects changes in a list and allows them to be replayed to a destination RDFChanges instance.
 * It supports both triples and quads based on the specified PatchStatementType.
 */
@ExperimentalApi
public final class JenaChangesCollector implements RDFChanges {

    private final List<JenaChangesItem> items = new ArrayList<>();
    private final PatchStatementType stType;

    /**
     * Creates a new JenaChangesCollector with the specified statement type.
     *
     * @param stType How to interpret the statements: TRIPLES or QUADS.
     */
    JenaChangesCollector(PatchStatementType stType) {
        this.stType = stType;
    }

    /**
     * Returns the number of changes collected.
     * @return The size of the collected changes.
     */
    public List<JenaChangesItem> getChanges() {
        return items;
    }

    /**
     * Returns the number of changes collected.
     * @return The size of the collected changes.
     */
    public int size() {
        return items.size();
    }

    /**
     * Replays the collected changes to the specified destination RDFChanges instance.
     * @param destination The RDFChanges instance to which the changes will be applied.
     * @param callStartFinish If true, calls start() and finish() on the destination.
     */
    public void replay(RDFChanges destination, boolean callStartFinish) {
        if (callStartFinish) destination.start();
        for (JenaChangesItem item : items) {
            item.applyTo(destination);
        }
        if (callStartFinish) destination.finish();
    }

    @Override
    public void header(String field, Node value) {
        items.add(new Header(field, value));
    }

    @Override
    public void add(Node g, Node s, Node p, Node o) {
        items.add(new Add(coerceGraph(g), s, p, o));
    }

    @Override
    public void delete(Node g, Node s, Node p, Node o) {
        items.add(new Delete(coerceGraph(g), s, p, o));
    }

    @Override
    public void addPrefix(Node gn, String prefix, String uriStr) {
        items.add(new AddPrefix(coerceGraph(gn), prefix, uriStr));
    }

    @Override
    public void deletePrefix(Node gn, String prefix) {
        items.add(new DeletePrefix(coerceGraph(gn), prefix));
    }

    @Override
    public void txnBegin() {
        items.add(TxnBegin.INSTANCE);
    }

    @Override
    public void txnCommit() {
        items.add(TxnCommit.INSTANCE);
    }

    @Override
    public void txnAbort() {
        items.add(TxnAbort.INSTANCE);
    }

    @Override
    public void segment() {
        items.add(Segment.INSTANCE);
    }

    @Override
    public void start() {}

    @Override
    public void finish() {}

    private Node coerceGraph(Node g) {
        if (g == null && stType == PatchStatementType.QUADS) {
            return Quad.defaultGraphNodeGenerated;
        } else if (stType == PatchStatementType.TRIPLES) {
            return null;
        } else {
            return g;
        }
    }

    /// Inner classes for JenaChangesItem

    public sealed interface JenaChangesItem {
        void applyTo(RDFChanges destination);
    }

    private record Header(String field, Node value) implements JenaChangesItem {
        @Override
        public void applyTo(RDFChanges destination) {
            destination.header(field, value);
        }
    }

    private record Add(Node g, Node s, Node p, Node o) implements JenaChangesItem {
        @Override
        public void applyTo(RDFChanges destination) {
            destination.add(g, s, p, o);
        }
    }

    private record Delete(Node g, Node s, Node p, Node o) implements JenaChangesItem {
        @Override
        public void applyTo(RDFChanges destination) {
            destination.delete(g, s, p, o);
        }
    }

    private record AddPrefix(Node gn, String prefix, String uriStr) implements JenaChangesItem {
        @Override
        public void applyTo(RDFChanges destination) {
            destination.addPrefix(gn, prefix, uriStr);
        }
    }

    private record DeletePrefix(Node gn, String prefix) implements JenaChangesItem {
        @Override
        public void applyTo(RDFChanges destination) {
            destination.deletePrefix(gn, prefix);
        }
    }

    private record TxnBegin() implements JenaChangesItem {
        public static final TxnBegin INSTANCE = new TxnBegin();

        @Override
        public void applyTo(RDFChanges destination) {
            destination.txnBegin();
        }
    }

    private record TxnCommit() implements JenaChangesItem {
        public static final TxnCommit INSTANCE = new TxnCommit();

        @Override
        public void applyTo(RDFChanges destination) {
            destination.txnCommit();
        }
    }

    private record TxnAbort() implements JenaChangesItem {
        public static final TxnAbort INSTANCE = new TxnAbort();

        @Override
        public void applyTo(RDFChanges destination) {
            destination.txnAbort();
        }
    }

    private record Segment() implements JenaChangesItem {
        public static final Segment INSTANCE = new Segment();

        @Override
        public void applyTo(RDFChanges destination) {
            destination.segment();
        }
    }
}
